pkg/provider/provider.go

```diff
@@ -272,82 +288,119 @@ func (p *Provider) buildServiceAndRoutersForTrafficSplit(t *topology.Topology, c
+			whitelistIndirect := p.buildWhitelistMiddlewareFromTrafficSplitIndirect(t, ts)
+			whitelistIndirectKey := getWhitelistMiddlewareKeyFromTrafficSplitIndirect(ts)
+			cfg.HTTP.Middlewares[whitelistIndirectKey] = whitelistIndirect
 
-			// If the ServiceTrafficSplit is a backend of at least one TrafficSplit we need an additional router with
-			// a whitelist middleware which whitelists based on the X-Forwarded-For header instead of on the RemoteAddr value.
-			if len(tsSvc.BackendOf) > 0 && p.config.ACL {
-				whitelistIndirect := p.buildWhitelistMiddlewareFromTrafficSplitIndirect(t, ts)
-				whitelistIndirectKey := getWhitelistMiddlewareKeyFromTrafficSplitIndirect(ts)
-				cfg.HTTP.Middlewares[whitelistIndirectKey] = whitelistIndirect
+			rule = buildHTTPRuleFromTrafficSplitIndirect(tsSvc)
+			rtrMiddlewaresindirect := addToSliceCopy(middlewares, whitelistIndirectKey)
+
+			indirectRtrKey := getRouterKeyFromTrafficSplitIndirect(ts, svcPort.Port)
+			cfg.HTTP.Routers[indirectRtrKey] = buildHTTPRouter(rule, entrypoint, rtrMiddlewaresindirect, svcKey, priorityTrafficTargetIndirect)
+		}
+	}
+}
+
+func (p *Provider) buildTCPServiceAndRoutersForTrafficSplit(cfg *dynamic.Configuration, tsKey topology.Key, ts *topology.TrafficSplit, tsSvc *topology.Service) {
+	tcpRule := buildTCPRouterRule()
 
-				rule = buildHTTPRuleFromTrafficSplitIndirect(tsSvc)
-				rtrMiddlewaresindirect := addToSliceCopy(middlewares, whitelistIndirectKey)
+	for _, svcPort := range tsSvc.Ports {
+		backendSvcs := make([]dynamic.TCPWRRService, len(ts.Backends))
 
-				indirectRtrKey := getRouterKeyFromTrafficSplitIndirect(ts, svcPort.Port)
-				cfg.HTTP.Routers[indirectRtrKey] = buildHTTPRouter(rule, entrypoint, rtrMiddlewaresindirect, svcKey, priorityTrafficTargetIndirect)
+		for i, backend := range ts.Backends {
+			backendSvcKey := getServiceKeyFromTrafficSplitBackend(ts, svcPort.Port, backend)
+			cfg.TCP.Services[backendSvcKey] = buildTCPSplitTrafficBackendService(backend, svcPort.TargetPort.IntVal)
+			backendSvcs[i] = dynamic.TCPWRRService{
+				Name:   backendSvcKey,
+				Weight: getIntRef(backend.Weight),
 			}
 		}
-	case k8s.ServiceTypeTCP:
-		tcpRule := buildTCPRouterRule()
-
-		for _, svcPort := range tsSvc.Ports {
-			backendSvcs := make([]dynamic.TCPWRRService, len(ts.Backends))
-
-			for i, backend := range ts.Backends {
-				backendSvcKey := getServiceKeyFromTrafficSplitBackend(ts, svcPort.Port, backend)
-				cfg.TCP.Services[backendSvcKey] = buildTCPSplitTrafficBackendService(backend, svcPort.TargetPort.IntVal)
-				backendSvcs[i] = dynamic.TCPWRRService{
-					Name:   backendSvcKey,
-					Weight: getIntRef(backend.Weight),
-				}
-			}
 
-			entrypoint, err := p.buildTCPEntrypoint(tsSvc, svcPort.Port)
-			if err != nil {
-				p.logger.Errorf("Unable to build TCP entrypoint for TrafficTarget %q and port %d: %v", tsKey, svcPort.Port, err)
-				continue
+		entrypoint, err := p.buildTCPEntrypoint(tsSvc, svcPort.Port)
```

Comment:
:thinking: I think there is a problem here. If the port is `UDP`, we are going to build backend services before we notice its not a TCP port. So we will end up with orphans backend services.
In order to fix that kind of bug, we must call `buildTCPEntrypoint` before doing anything. Another option would be to be more explicit on which  protocol we are interested in. This could be done by checking the `svcPort.Protocol != "TCP"`  before iterating on `ts.Backends`. This kind of issue is not obvious to detect. I suggest we filter ports in these loops everywhere.

Reply:
You're right! I think we have to be explicit because we're creating false errors in my point of view and that should be addressed in another PR/Issue. For now, as you said I've called `buildTCPEntrypoint` and `buildUDPEntrypoint` before doing anything.